fastwsgi/server.c

```diff
@@ -462,51 +489,222 @@ int main()
     uv__socket_sockopt((uv_handle_t*)&g_srv.server, so_reuseport, &enabled);
 #endif
 
-    int err = uv_tcp_bind(&g_srv.server, (const struct sockaddr*)&addr, 0);
+    int err = uv_tcp_bind(&g_srv.server, &addr.addr, 0);
     if (err) {
         LOGe("Bind error %s\n", uv_strerror(err));
-        return 1;
+        hr = -5;
+        goto fin;
     }
-
     err = uv_listen((uv_stream_t*)&g_srv.server, g_srv.backlog, connection_cb);
     if (err) {
         LOGe("Listen error %s\n", uv_strerror(err));
-        return 1;
+        hr = -6;
+        goto fin;
+    }    
+    if (g_srv.hook_sigint > 0) {
+        uv_signal_init(g_srv.loop, &g_srv.signal);
+        uv_signal_start(&g_srv.signal, signal_handler, SIGINT);
     }
+    g_srv_inited = 1;
+    hr = 0;
 
-    uv_signal_t signal;
-    uv_signal_init(g_srv.loop, &signal);
-    uv_signal_start(&signal, signal_handler, SIGINT);
+fin:
+    if (hr) {
+        if (g_srv.signal.signal_cb)
+            uv_signal_stop(&g_srv.signal);
 
-    uv_run(g_srv.loop, UV_RUN_DEFAULT);
-    uv_loop_close(g_srv.loop);
-    free(g_srv.loop);
-    return 0;
+        if (hr <= -5)
+            uv_close((uv_handle_t *)&g_srv, NULL);
+
+        if (g_srv.loop)
+            uv_loop_close(g_srv.loop);
+
+        memset(&g_srv, 0, sizeof(g_srv));
+    }    
+    return hr;
 }
 
-PyObject * run_server(PyObject * self, PyObject * args)
+PyObject * init_server(PyObject * Py_UNUSED(self), PyObject * server)
 {
     int64_t rv;
+
+    if (g_srv_inited) {
+        PyErr_Format(PyExc_Exception, "server already inited");
+        return PyLong_FromLong(-1000);
+    }
     memset(&g_srv, 0, sizeof(g_srv));
-    int log_level = 0;
-    PyArg_ParseTuple(args, "Osiii", &g_srv.wsgi_app, &g_srv.host, &g_srv.port, &g_srv.backlog, &log_level);
-    set_log_level(log_level);
+    g_srv.pysrv = server;
+
+    int64_t loglevel = get_obj_attr_int(server, "loglevel");
+    if (loglevel == LLONG_MIN) {
+        PyErr_Format(PyExc_ValueError, "Option loglevel not defined");
+        return PyLong_FromLong(-1010);
+    }
+    set_log_level((int)loglevel);
+
+    g_srv.wsgi_app = PyObject_GetAttrString(server, "app");
+    Py_XDECREF(g_srv.wsgi_app);
```

Comment:
I see the assign then decref pattern is also used here. If this is valid, I'm happy to approve and merge this. 
Just want to know if its a valid way of tracking ref counts and doesn't cause errors if the variable is used after the decref...

Reply:
Objects `fastwsgi.server` and `g_srv` partially duplicate attributes/properties. But these objects must be destroyed separately. It's just that some properties are copied (rather than inherited) from object `fastwsgi.server` to object `g_srv`.